modules/ctl/src/main/scala/free/interp.scala

```diff
@@ -0,0 +1,119 @@
+package gem.ctl
+package free
+
+import scalaz._, Scalaz._
+import scalaz.effect._
+
+import gem.ctl.low.io._
+import gem.ctl.free.ctl._
+
+/** An interpreter for the control language. */
+object interpreter {
+
+  /** Configuration for the interpreter. */
+  trait Config {
+    def verbose: Boolean
+    def server:  Server
+  }
+
+  case class InterpreterState(indentation: Int, machineHostCache: Map[Host.Machine, String]) {
+    def indent  = copy(indentation = indentation + 1)
+    def outdent = copy(indentation = indentation - 1)
+  }
+  object InterpreterState {
+    val initial = InterpreterState(0, Map.empty)
+  }
+
+  /**
+   * Construct an interpreter of `CtlIO` given a `Config` and an `IORef` for the initial state. We
+   * carry our state in an `IORef` because it needs to be visible from multiple threads.
+   */
+  def interpreter(c: Config, state: IORef[InterpreterState]) = λ[CtlOp ~> EitherT[IO, Int, ?]] {
```

Comment:
It is probably very standard if you are very deep into the FP stack but becomes very obscure if you are not. So the question is how do we get there. At least I have a very vague idea of what a natural transformation is so having more syntax on top doesn't make it simpler